src/smtp_server.h

```diff
@@ -76,6 +81,35 @@ struct smtp_cmd_tree {
 	struct list_head hdlrs;
 };
 
+/**
+ * Define SMTP preprocess handlers
+ */
+int smtp_hdlr_init(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_auth(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_alou(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_alop(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_aplp(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_ehlo(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_data(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_mail(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_rcpt(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_rset(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+int smtp_hdlr_quit(struct smtp_server_context *ctx, const char *cmd, const char *arg, bfd_t *stream);
+
+/**
+ * SMTP command structure
+ *
+ * Used by a SMTP command to maintain the C preprocess handler
+ * and JS stub function
+ */
+ struct smtp_cmd_hdlr {
+	const char cmd_name[4];
+	int (*smtp_preprocess_hdlr)(struct smtp_server_context *, const char *, const char *, bfd_t*);
```

Comment:
For the smtp_preprocess_hdlr field, you can (re)use smtp_cmd_hdlr_t which is defined above. The prototype is identical.

You should delete struct smtp_cmd_hdlr_list. It's no longer used and it's confusing.